src/librustc_mir/borrow_check/path_utils.rs

```diff
@@ -138,24 +138,27 @@ pub(super) fn is_active<'tcx>(
 
 /// Determines if a given borrow is borrowing local data
 /// This is called for all Yield statements on movable generators
-pub(super) fn borrow_of_local_data<'tcx>(place: &Place<'tcx>) -> bool {
-    match place {
-        Place::Promoted(_) |
-        Place::Static(..) => false,
-        Place::Local(..) => true,
-        Place::Projection(box proj) => {
-            match proj.elem {
-                // Reborrow of already borrowed data is ignored
-                // Any errors will be caught on the initial borrow
-                ProjectionElem::Deref => false,
+pub(super) fn borrow_of_local_data<'a, 'tcx, 'gcx: 'tcx>(
+    tcx: TyCtxt<'a, 'gcx, 'tcx>,
+    place: &Place<'tcx>
+) -> bool {
+    if let Some((base_place, projection)) = place.split_projection(tcx) {
+        match projection {// Reborrow of already borrowed data is ignored
+            // Any errors will be caught on the initial borrow
+            ProjectionElem::Deref => false,
 
-                // For interior references and downcasts, find out if the base is local
-                ProjectionElem::Field(..)
-                    | ProjectionElem::Index(..)
-                    | ProjectionElem::ConstantIndex { .. }
-                | ProjectionElem::Subslice { .. }
-                | ProjectionElem::Downcast(..) => borrow_of_local_data(&proj.base),
-            }
+            // For interior references and downcasts, find out if the base is local
+            ProjectionElem::Field(..)
+            | ProjectionElem::Index(..)
+            | ProjectionElem::ConstantIndex { .. }
+            | ProjectionElem::Subslice { .. }
+            | ProjectionElem::Downcast(..) => borrow_of_local_data(tcx, &base_place),
+        }
+    } else {
+        match place.base {
+            PlaceBase::Promoted(_)
+            | PlaceBase::Static(..) => false,
+            PlaceBase::Local(..) => true,
```

Comment:
@eddyb there are plenty of such `pattern-match`s on `old_place`, I'm a bit uncertain this translation, is this correct?

Reply:
`borrow_of_local_data` is a recursive function, which means you can make it iterate over the `elems`.

Furthermore, in this case specifically, if you made it iterative, you'd notice that the `elems` don't change *at all* whether the function returns `true` or `false`, it just needs to look at the base.